compiler/vm/vmdef.nim

```diff
@@ -269,22 +269,41 @@ type
     fnc*: VmFunctionPtr
     env*: HeapSlotHandle
 
-  # XXX: once first class openArray support is available during bootstrapping,
-  #      `MemRegionPtr` could be turned into `openArray[byte]`
-  MemRegionPtr* = object
-    ## A pointer to a region of memory
-    p: ptr UncheckedArray[byte]
-    l: int
+  VmMemPointer* = distinct ptr UncheckedArray[byte]
+    ## A pointer into a memory region managed by the VM (i.e. guest memory)
   VmMemoryRegion* = openArray[byte]
 
+  CellId* = int
+  CellPtr* = distinct ptr UncheckedArray[byte]
+    ## A pointer that either is nil or points to the start of a VM memory cell
+
   LocHandle* = object
-    ## A handle to a ``location``
-    h*: MemRegionPtr
-    typ* {.cursor.}: PVmType # ownership of `typ` is not needed
+    ## A handle to a VM memory location. Since looking up the cell the location
+    ## is part of is a very common operation (due to access checks), the cell's
+    ## ID is also stored here.
+    cell*: CellId
+      ## the ID of the cell that owns the location
+    p*: VmMemPointer
+      ## the memory address of the reference location location, in host address
+      ## space
+    typ* {.cursor.}: PVmType
+      ## the type that the handles views the location as
+
+  VmSlice* = object
```

Comment:
I wouldn't change it now, and I'm not entirely sold on the name, but `VmMemHostSlice` might be indicative that this is host only.

Reply:
I'm also not happy with `VmSlice` -- the name should be reserved for the soon-to-be in-memory representation for guest `openArray`s (similar to how `VmSeq` is used as the name for the guest `seq` representation). 

To me `VmMemHostSlice` sounds like it's a slice into host memory, which, while technically correct (the host and guest share the same address space), doesn't match the intention, i.e. that it's a guest memory slice.